src/blockfilter.cpp

```diff
@@ -153,3 +153,47 @@ GCSFilter::GCSFilter(uint64_t siphash_k0, uint64_t siphash_k1, uint8_t P,
 
     bitwriter.Flush();
 }
+
+bool GCSFilter::MatchInternal(const uint64_t* element_hashes, size_t size) const
+{
+    CVectorReader stream(GCS_SER_TYPE, GCS_SER_VERSION, m_encoded, 0);
+
+    // Seek forward by size of N
+    uint64_t N = ReadCompactSize(stream);
+    assert(N == m_N);
+
+    BitStreamReader<CVectorReader> bitreader(stream);
+
+    uint64_t value = 0;
+    int hashes_index = 0;
+    for (uint64_t i = 0; i < m_N; i++) {
+        uint64_t delta = GolombRiceDecode(bitreader, m_P);
+        value += delta;
+
+        while (true) {
+            if (hashes_index == size) {
+                return false;
+            } else if (element_hashes[hashes_index] == value) {
+                return true;
+            } else if (element_hashes[hashes_index] > value) {
+                break;
+            }
+
+            hashes_index++;
+        }
+    }
+
+    return false;
+}
+
+bool GCSFilter::Match(const GCSFilter::Element& element) const
+{
+    uint64_t query = HashToRange(element);
+    return MatchInternal(&query, 1);
+}
+
+bool GCSFilter::MatchAny(const std::set<Element>& elements) const
+{
+    const std::vector<uint64_t>&& queries = BuildHashedSet(elements);
```

Comment:
I don't think the reference type is useful here; copy elision should apply, and otherwise at worst a move will occur.